File: offhand_test.py
import asyncore
import os
import socket
import subprocess
import sys
import threading
import time

import offhand

messages = [
	# TestSequence
	["\xf0\x00\x0d"],
	[],
	["", "\xde\xad\xbe\xef\xca\xfe\xba", "", "", "\xbe", ""],

	# TestParallel
	["\xff\xee\xdd"],
	[],
	["", "\xcc\xcc\xcc\xbb\xbb\xbb\xaa", "", "", "\xaa", ""],
]

class Puller(offhand.AsynConnectPuller):

	class Node(offhand.AsynConnectPuller.Node):
		socket_family = socket.AF_UNIX

	def handle_pull(self, node, message):
		print self, node, "message", message
		messages.remove(message)

def test(done):
	os.chdir(".test/sockets")

	with Puller() as p1, Puller() as p2:
		pullers = p1, p2

		for i in xrange(3):
			for p in pullers:
				p.connect(str(i))

		timeout   = 0.1
		last_time = time.time()

		while True:
			t = timeout - (time.time() - last_time)
			if t <= 0:
				for p in pullers:
					p.reconnect()

				last_time = time.time()
				t = timeout

			if asyncore.socket_map:
				asyncore.loop(timeout=t, use_poll=True, count=1)
			elif done[0]:
				return
			else:
				time.sleep(t)

def main():
	done = [False]

	t = threading.Thread(target=test, args=(done,))
	t.daemon = True

	p = subprocess.Popen(["./offhand.test", "-test.v=true"])
	try:
		t.start()
	finally:
		status = p.wait()
		if status == 0:
			print "go test exited"
		else:
			print >>sys.stderr, "go test exited with %r" % status

	done[0] = True
	t.join(2)
	if t.is_alive():
		print >>sys.stderr, "python test exit timeout"
	else:
		print "python test exited"

		if messages:
			print >>sys.stderr, "remaining messages:"
			for m in messages:
				print >>sys.stderr, "  %r" % m

if __name__ == "__main__":
	try:
		main()
	except KeyboardInterrupt:
		print

File: offhand_test.go
package offhand

import (
	"fmt"
	"net"
	"testing"
)

func new_listener(t *testing.T, n int) (l net.Listener) {
	l, err := net.Listen("unix", fmt.Sprintf(".test/sockets/%d", n))
	if err != nil {
		t.Fatalf("listen %d: %s", n, err.Error())
	}
	return
}

func new_pusher(t *testing.T, n int) Pusher {
	return NewListenPusher(new_listener(t, n))
}

func send(t *testing.T, p Pusher, n int, message [][]byte) {
	err := p.SendMultipart(message)
	if err != nil {
		t.Errorf("send %d: %s", n, err.Error())
	}
}

func close_with_stats(t *testing.T, n int, p Pusher) {
	p.Close()
	t.Log("pusher", n, "stats:", p.Stats().String())
}

func push(t *testing.T, exit func(), n int, message [][]byte) {
	p := new_pusher(t, n)
	send(t, p, n, message)
	close_with_stats(t, n, p)
	exit()
}

func TestSequence(t *testing.T) {
	p := new_pusher(t, 0)

	send(t, p, 0, [][]byte{ []byte{ 0xf0, 0x00, 0x0d } })
	send(t, p, 0, [][]byte{})
	send(t, p, 0, [][]byte{ []byte{}, []byte{ 0xde, 0xad, 0xbe, 0xef, 0xca, 0xfe, 0xba }, []byte{}, []byte{}, []byte{ 0xbe }, []byte{} })

	close_with_stats(t, 0, p)

	if p.SendMultipart([][]byte{}) == nil {
		t.Error("send succeeded after close")
	}

	close_with_stats(t, 0, new_pusher(t, 0))
	close_with_stats(t, 1, new_pusher(t, 1))
}

func TestParallel(t *testing.T) {
	done := make(chan bool)
	exit := func() { done<- true }

	go push(t, exit, 0, [][]byte{ []byte{ 0xff, 0xee, 0xdd } })
	go push(t, exit, 1, [][]byte{})
	go push(t, exit, 2, [][]byte{ []byte{}, []byte{ 0xcc, 0xcc, 0xcc, 0xbb, 0xbb, 0xbb, 0xaa }, []byte{}, []byte{}, []byte{ 0xaa }, []byte{} })

	<-done
	<-done
	<-done

	p0 := new_pusher(t, 0)
	p1 := new_pusher(t, 1)
	p2 := new_pusher(t, 2)
	close_with_stats(t, 2, p2)
	close_with_stats(t, 0, p0)
	close_with_stats(t, 1, p1)
}
